Make DeviceMesh opaque

angelayi opened the pull request:
By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #169867



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @ezyang

Differential Revision: [D94288113](https://our.internmc.facebook.com/intern/diff/D94288113)

angelayi committed:
[opaque_obj] Try out dtensor tests

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Try out dtensor tests"




cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Try out dtensor tests"




cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Try out dtensor tests"




cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Try out dtensor tests"

By marking DTensorSpec as being a value-type opaque object, the graph looks like:
```
def forward(self, b_buffer, x):
    _assert_tensor_metadata_default = torch.ops.aten._assert_tensor_metadata.default(x, dtype = torch.float64, device = device(type='cpu'), layout = torch.strided);  _assert_tensor_metadata_default = None
    to = torch.ops.aten.to.dtype_layout(x, dtype = torch.float64, layout = torch.strided, device = device(type='{self.device_type}'));  x = None
    view_as = torch.ops.aten.view_as.default(to, to);  to = None
    dtensor___init__0 = self.dtensor___init__0
    dtensor_const_func_spec0 = self.dtensor_const_func_spec0
    flat_apply = torch.ops.higher_order.flat_apply(dtensor_const_func_spec0, dtensor___init__0, view_as, DTensorSpec(mesh=DeviceMesh('{self.device_type}', [0, 1]), placements=(Shard(dim=0),), tensor_meta=TensorMeta(shape=torch.Size([8, 4]), stride=(4, 1), dtype=torch.float64), shard_order=(ShardOrderEntry(tensor_dim=0, mesh_dims=(0,)),)), False);  dtensor_const_func_spec0 = dtensor___init__0 = view_as = None
    add = torch.ops.aten.add.Tensor(b_buffer, flat_apply);  b_buffer = flat_apply = None
    access_subclass_inner_tensor_default_4 = torch.ops.export.access_subclass_inner_tensor.default(add, '_local_tensor');  add = None
    view_as_1 = torch.ops.aten.view_as.default(access_subclass_inner_tensor_default_4, access_subclass_inner_tensor_default_4);  access_subclass_inner_tensor_default_4 = None
    return (view_as_1,)""",  # noqa: B950
        )
```




cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

ezyang commented on torch/distributed/tensor/_dtensor_spec.py:
```diff
@@ -410,6 +433,38 @@ def _check_equals(self, other: object, skip_shapes: bool = False) -> bool:
 
+    def __fx_eval__(self):
+        """
+        Returns FX-evaluable repr and required globals for DTensorSpec.
```
In fact I don't want any of these in this file for custom ops, I think

ezyang commented on test/distributed/tensor/test_dtensor_compile.py:
```diff
@@ -242,7 +261,7 @@ def forward(self, b_buffer, x):
     dtensor___init__0 = self.dtensor___init__0
     dtensor_const_func_spec0 = self.dtensor_const_func_spec0
-    flat_apply = torch.ops.higher_order.flat_apply(dtensor_const_func_spec0, dtensor___init__0, view_as, False);  dtensor_const_func_spec0 = dtensor___init__0 = view_as = None
+    flat_apply = torch.ops.higher_order.flat_apply(dtensor_const_func_spec0, dtensor___init__0, view_as, torch.distributed.tensor._dtensor_spec.DTensorSpec(mesh=torch.distributed.device_mesh.DeviceMesh('{self.device_type}', [0, 1]), placements=(torch.distributed.tensor.placement_types.Shard(dim=0),), tensor_meta=torch.distributed.tensor._dtensor_spec.TensorMeta(shape=torch.Size([8, 4]), stride=(4 ...
```
I think this test is just bad. I want to see something with a signature much more like DTensor.from_local in the graph here. I think strict=True will do this?

ezyang commented on torch/distributed/device_mesh.py:
```diff
@@ -206,6 +206,16 @@ def __init__(
+                    from torch._subclasses.fake_tensor import unset_fake_temporarily
+
+                    with unset_fake_temporarily():
+                        self._concrete_mesh = mesh.detach().cpu().tolist()
```
This is probably a bad idea though I'm not sure what to propose instead

angelayi committed:
Update on "[opaque_obj] Try out dtensor tests"

By marking DTensorSpec as being a value-type opaque object, the graph looks like:
```python
def forward(self, b_buffer, x):
    _assert_tensor_metadata_default = torch.ops.aten._assert_tensor_metadata.default(x, dtype = torch.float64, device = device(type='cpu'), layout = torch.strided);  _assert_tensor_metadata_default = None
    to = torch.ops.aten.to.dtype_layout(x, dtype = torch.float64, layout = torch.strided, device = device(type='cuda'));  x = None
    view_as = torch.ops.aten.view_as.default(to, to);  to = None
    dtensor___init__0 = self.dtensor___init__0
    dtensor_const_func_spec0 = self.dtensor_const_func_spec0
    flat_apply = torch.ops.higher_order.flat_apply(
        dtensor_const_func_spec0, 
        dtensor___init__0, 
        view_as,
        torch.distributed.tensor._dtensor_spec.DTensorSpec(
            mesh=torch.distributed.device_mesh.DeviceMesh('cuda', [0, 1]), 
            placements=(torch.distributed.tensor.placement_types.Shard(dim=0),), 
            tensor_meta=torch.distributed.tensor._dtensor_spec.TensorMeta(shape=torch.Size([8, 4]), stride=(4, 1), dtype=torch.float64), 
            shard_order=(torch.distributed.tensor._dtensor_spec.ShardOrderEntry(tensor_dim=0, mesh_dims=(0,)),)), 
        False);
    add = torch.ops.aten.add.Tensor(b_buffer, flat_apply);  b_buffer = flat_apply = None
    access_subclass_inner_tensor_default_4 = torch.ops.export.access_subclass_inner_tensor.default(add, '_local_tensor');  add = None
    view_as_1 = torch.ops.aten.view_as.default(access_subclass_inner_tensor_default_4, access_subclass_inner_tensor_default_4);  access_subclass_inner_tensor_default_4 = None
    return (view_as_1,)
```




cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Try out dtensor tests"

By marking DTensorSpec as being a value-type opaque object, the graph looks like:
```python
def forward(self, b_buffer, x):
    _assert_tensor_metadata_default = torch.ops.aten._assert_tensor_metadata.default(x, dtype = torch.float64, device = device(type='cpu'), layout = torch.strided);  _assert_tensor_metadata_default = None
    to = torch.ops.aten.to.dtype_layout(x, dtype = torch.float64, layout = torch.strided, device = device(type='cuda'));  x = None
    view_as = torch.ops.aten.view_as.default(to, to);  to = None
    dtensor___init__0 = self.dtensor___init__0
    dtensor_const_func_spec0 = self.dtensor_const_func_spec0
    flat_apply = torch.ops.higher_order.flat_apply(
        dtensor_const_func_spec0, 
        dtensor___init__0, 
        view_as,
        torch.distributed.tensor._dtensor_spec.DTensorSpec(
            mesh=torch.distributed.device_mesh.DeviceMesh('cuda', [0, 1]), 
            placements=(torch.distributed.tensor.placement_types.Shard(dim=0),), 
            tensor_meta=torch.distributed.tensor._dtensor_spec.TensorMeta(shape=torch.Size([8, 4]), stride=(4, 1), dtype=torch.float64), 
            shard_order=(torch.distributed.tensor._dtensor_spec.ShardOrderEntry(tensor_dim=0, mesh_dims=(0,)),)), 
        False);
    add = torch.ops.aten.add.Tensor(b_buffer, flat_apply);  b_buffer = flat_apply = None
    access_subclass_inner_tensor_default_4 = torch.ops.export.access_subclass_inner_tensor.default(add, '_local_tensor');  add = None
    view_as_1 = torch.ops.aten.view_as.default(access_subclass_inner_tensor_default_4, access_subclass_inner_tensor_default_4);  access_subclass_inner_tensor_default_4 = None
    return (view_as_1,)
```




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Trace dtensor functions through opaque objects infra"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_self_buffers_buffer_ : torch.distributed.tensor.DTensor, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_self_buffers_buffer_ = L_self_buffers_buffer_
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    from_local = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = l_mesh_ = None
    inter = l_self_buffers_buffer_ + from_local;  l_self_buffers_buffer_ = from_local = None
    to_local = inter.to_local();  inter = None
    return (to_local,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local/to_local directly (since they are TorchInGraphFunctionVariable). Note: this doesn't seem to solve the caching problem though, since `torch.distributed.tensor._api.from_local` seems to be unsupported.

The AOTAutograd graph decomposes the from_local/to_local operations and looks like:
```
def forward(self, arg0_1, arg1_1, arg2_1):
    _to_copy = torch.ops.aten._to_copy.default(arg1_1, dtype = torch.float64, layout = torch.strided, device = device(type='cuda', index=0));  arg1_1 = None
    view = torch.ops.aten.view.default(_to_copy, [4, 4]);  _to_copy = None
    add = torch.ops.aten.add.Tensor(arg0_1, view);  arg0_1 = view = None
    view_1 = torch.ops.aten.view.default(add, [4, 4]);  add = None
    return (view_1,)
```




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Trace dtensor functions through opaque objects infra"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_self_buffers_buffer_ : torch.distributed.tensor.DTensor, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_self_buffers_buffer_ = L_self_buffers_buffer_
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    from_local = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = l_mesh_ = None
    inter = l_self_buffers_buffer_ + from_local;  l_self_buffers_buffer_ = from_local = None
    to_local = inter.to_local();  inter = None
    return (to_local,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local/to_local directly (since they are TorchInGraphFunctionVariable). Note: this doesn't seem to solve the caching problem though, since `torch.distributed.tensor._api.from_local` seems to be unsupported.

The AOTAutograd graph decomposes the from_local/to_local operations and looks like:
```
def forward(self, arg0_1, arg1_1, arg2_1):
    _to_copy = torch.ops.aten._to_copy.default(arg1_1, dtype = torch.float64, layout = torch.strided, device = device(type='cuda', index=0));  arg1_1 = None
    view = torch.ops.aten.view.default(_to_copy, [4, 4]);  _to_copy = None
    add = torch.ops.aten.add.Tensor(arg0_1, view);  arg0_1 = view = None
    view_1 = torch.ops.aten.view.default(add, [4, 4]);  add = None
    return (view_1,)
```




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Trace dtensor functions through opaque objects infra"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_self_buffers_buffer_ : torch.distributed.tensor.DTensor, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_self_buffers_buffer_ = L_self_buffers_buffer_
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    from_local = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = l_mesh_ = None
    inter = l_self_buffers_buffer_ + from_local;  l_self_buffers_buffer_ = from_local = None
    to_local = inter.to_local();  inter = None
    return (to_local,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local/to_local directly (since they are TorchInGraphFunctionVariable). Note: this doesn't seem to solve the caching problem though, since `torch.distributed.tensor._api.from_local` seems to be unsupported.

The AOTAutograd graph decomposes the from_local/to_local operations and looks like:
```
def forward(self, arg0_1, arg1_1, arg2_1):
    _to_copy = torch.ops.aten._to_copy.default(arg1_1, dtype = torch.float64, layout = torch.strided, device = device(type='cuda', index=0));  arg1_1 = None
    view = torch.ops.aten.view.default(_to_copy, [4, 4]);  _to_copy = None
    add = torch.ops.aten.add.Tensor(arg0_1, view);  arg0_1 = view = None
    view_1 = torch.ops.aten.view.default(add, [4, 4]);  add = None
    return (view_1,)
```




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Trace dtensor functions through opaque objects infra"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_self_buffers_buffer_ : torch.distributed.tensor.DTensor, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_self_buffers_buffer_ = L_self_buffers_buffer_
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    from_local = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = l_mesh_ = None
    inter = l_self_buffers_buffer_ + from_local;  l_self_buffers_buffer_ = from_local = None
    to_local = inter.to_local();  inter = None
    return (to_local,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local/to_local directly (since they are TorchInGraphFunctionVariable). Note: this doesn't seem to solve the caching problem though, since `torch.distributed.tensor._api.from_local` seems to be unsupported.

The AOTAutograd graph decomposes the from_local/to_local operations and looks like:
```
def forward(self, arg0_1, arg1_1, arg2_1):
    _to_copy = torch.ops.aten._to_copy.default(arg1_1, dtype = torch.float64, layout = torch.strided, device = device(type='cuda', index=0));  arg1_1 = None
    view = torch.ops.aten.view.default(_to_copy, [4, 4]);  _to_copy = None
    add = torch.ops.aten.add.Tensor(arg0_1, view);  arg0_1 = view = None
    view_1 = torch.ops.aten.view.default(add, [4, 4]);  add = None
    return (view_1,)
```




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

ezyang commented on torch/distributed/device_mesh.py:
```diff
@@ -1378,3 +1380,55 @@ def init_device_mesh(
+            obj._layout,
+            obj._device_type,
+            obj._mesh_dim_names,
+            obj._thread_id,
```
Also concerned about perf overhead of the guard here.

ezyang replied:
Claude thinks perf overhead is fine

```
  Attributes and their types:

  | Attribute         | Type                           | Equality Cost                                             |
  |-------------------|--------------------------------|-----------------------------------------------------------|
  | _flatten_rank_map | tuple[int, ...]                | O(n) where n = number of ranks in mesh                    |
  | _layout           | _MeshLayout (frozen dataclass) | Very cheap - compares two IntTuple fields (shape, stride) |
  | _device_type      | str                            | Very cheap - typically "cuda" or "cpu"                    |
  | _mesh_dim_names   | tuple[str, ...] | None         | Cheap - typically 1-3 short dimension names               |
  | _thread_id        | None or int                    | Trivial                                                   |

  The main cost driver is _flatten_rank_map:
  - It's defined at line 264 as tuple(self._rank_map.tolist())
  - This is a tuple of all rank indices in the mesh
  - For a typical 2D mesh of 8 GPUs, this is just 8 integers → negligible
  - For large meshes (e.g., 1024 GPUs), it's 1024 integer comparisons → still quite fast

  Overall assessment: Cheap

  The guard evaluation is essentially what DeviceMesh.__eq__ already does (lines 539-550), and the values are:
  1. Pre-computed at mesh creation time (no runtime conversion cost)
  2. Small Python primitives (tuples of ints/strings)
  3. Already being used for __hash__ computation

  For typical distributed training scenarios (8-256 GPUs), you're looking at microseconds. Even for very large meshes, this is a cheap operation compared to the graph compilation overhead that guards are protecting against.
```

ezyang reviewed (approved):
Beautiful, thanks!

bdhirsh commented on torch/distributed/device_mesh.py:
```diff
@@ -1378,3 +1380,55 @@ def init_device_mesh(
+    register_opaque_type(
+        DeviceMesh,
+        typ="reference",
+        guard_fn=lambda obj: [
```
(totally does not need to block this PR, I'm just thinking about overhead issues)

bdhirsh replied:
oh ha from Ed's comment above, claude thinks it's fine 😛

bdhirsh commented on torch/distributed/device_mesh.py:
```diff
@@ -1378,3 +1380,55 @@ def init_device_mesh(
+            obj._mesh_dim_names,
+            obj._thread_id,
+        ],
+        members={
```
is there a convenient way of erroring loudly if someone adds a new DeviceMesh field and forgets to update this mapping? (maybe a unit test that walks DeviceMesh attribute names and matches them to this dict?)

angelayi replied:
If a new DeviceMesh field gets added and they don't update this mapping, when they try to use the field within some function/code that is being torch.compiled, we'll error saying "hey you cant access this -- you have to add it to the members list" . Is that ok?

zou3519 replied:
To clarify, is it going to graph break or is it a loud error? And which is preferred?

angelayi committed:
Update on "Trace dtensor functions through opaque objects infra"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zou3519 commented on test/distributed/tensor/test_dtensor_compile.py:
```diff
@@ -177,40 +183,91 @@ def forward(self, x):
+    l_self_buffers_buffer_ = L_self_buffers_buffer_
+    l_x_ = L_x_
+    l_mesh_ = L_mesh_
+    from_local = torch.distributed.tensor._api.DTensor.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = l_mesh_ = None
```
@bdhirsh this is the line we were discussing before. Do you care if this function is allowed into the dynamo graph (and we tell AOTAutogradCache that it is cache-safe) or if we should turn it into a custom operator?

angelayi committed:
Update on "Trace dtensor functions through opaque objects infra"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Trace dtensor functions through opaque objects infra"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Trace dtensor functions through opaque objects infra"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make DeviceMesh opaque"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make DeviceMesh opaque"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make DeviceMesh opaque"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make DeviceMesh opaque"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela ezyang

[ghstack-poisoned]

angelayi committed:
Update on "Make DeviceMesh opaque"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela ezyang

[ghstack-poisoned]

angelayi committed:
Update on "Make DeviceMesh opaque"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela ezyang

[ghstack-poisoned]

angelayi committed:
Update on "Make DeviceMesh opaque"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela ezyang

[ghstack-poisoned]

angelayi committed:
Update on "Make DeviceMesh opaque"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela ezyang

[ghstack-poisoned]

angelayi committed:
Update on "Make DeviceMesh opaque"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela ezyang

[ghstack-poisoned]

angelayi committed:
Update on "Make DeviceMesh opaque"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela ezyang

[ghstack-poisoned]

aorenste commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1573,6 +1573,7 @@ def method___contains__(
         )
         return result.call_method(tx, "item", [], {})
 
+<<<<<<< HEAD
```
This looks like a messed-up merge...

angelayi committed:
Update on "Make DeviceMesh opaque"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela ezyang

[ghstack-poisoned]

angelayi committed:
Update on "Make DeviceMesh opaque"

By marking DTensorSpec as being a value-type opaque object, the dynamo graph now looks like:
```python
def forward(self, L_x_ : torch.Tensor, L_mesh_ : torch.distributed.device_mesh.DeviceMesh):
    l_x_ = L_x_
    l_mesh_ = L_mesh_
    dt = torch.distributed.tensor._api.from_local(l_x_, l_mesh_, [torch.distributed.tensor.placement_types.Shard(dim=0)], run_check = False);  l_x_ = None
    redistribute = dt.redistribute(device_mesh = l_mesh_, placements = [torch.distributed.tensor.placement_types.Replicate()]);  dt = l_mesh_ = None
    to_local = redistribute.to_local();  redistribute = None
    add = to_local + 2;  to_local = None
    return (add,)
```
It takes in the DeviceMesh as an input (since it is marked as reference-type opaque object), and calls from_local directly (since it is marked as TorchInGraphFunctionVariable), and to_local/redistribute as call_methods on the tensors

The AOTAutograd graph decomposes the from_local/to_local/redistribute operations and looks like:
```
def forward(self, arg0_1, arg1_1):
    _to_copy = torch.ops.aten._to_copy.default(arg0_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  arg0_1 = None
    view = torch.ops.aten.view.default(_to_copy, [1]);  _to_copy = None
    all_gather_into_tensor = torch.ops._c10d_functional.all_gather_into_tensor.default(view, 2, '0');  view = None
    wait_tensor = torch.ops._c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
    view_1 = torch.ops.aten.view.default(wait_tensor, [2]);  wait_tensor = None
    add = torch.ops.aten.add.Tensor(view_1, 2);  view_1 = None
    return (add,)
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela ezyang

[ghstack-poisoned]
